src/extension-consumer/src/consume/mod.rs

```diff
@@ -74,6 +92,160 @@ impl ConsumeLogOpt {
         }
     }
+
+    /// Consume records according to the given configuration.
+    pub async fn consume_records(&self, consumer: PartitionConsumer) -> Result<(), ConsumerError> {
+        debug!("starting fetch loop: {:#?}", self);
+        self.init_ctrlc()?;
+        let offset = self.calculate_offset()?;
+
+        let consume_config = {
+            let mut config = ConsumerConfig::default();
+            if let Some(max_bytes) = self.max_bytes {
+                config = config.with_max_bytes(max_bytes);
+            }
+            config
+        };
+
+        if self.disable_continuous {
+            self.consume_records_batch(&consumer, offset, consume_config)
+                .await?;
+        } else {
+            self.consume_records_stream(&consumer, offset, consume_config)
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    /// Consume records in a single batch, then exit
+    async fn consume_records_batch(
+        &self,
+        consumer: &PartitionConsumer,
+        offset: Offset,
+        config: ConsumerConfig,
+    ) -> Result<(), ConsumerError> {
+        let response = consumer.fetch_with_config(offset, config).await?;
+
+        debug!(
+            "got a single response: LSO: {} batches: {}",
+            response.log_start_offset,
+            response.records.batches.len(),
+        );
+
+        for batch in response.records.batches.iter() {
+            for record in batch.records().iter() {
+                let key = record.key.as_ref().map(|it| it.as_ref());
+                self.print_record(key, record.value.as_ref())?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Consume records as a stream, waiting for new records to arrive
+    async fn consume_records_stream(
+        &self,
+        consumer: &PartitionConsumer,
+        offset: Offset,
+        config: ConsumerConfig,
+    ) -> Result<(), ConsumerError> {
+        let mut stream = consumer.stream_with_config(offset, config).await?;
+
+        while let Some(result) = stream.next().await {
+            let result: Result<Record, _> = result;
+            let record = match result {
+                Ok(record) => record,
+                Err(FluvioError::ApiError(ApiError::Code(code, _))) => {
+                    println!(
+                        "topic '{}/{}': {}",
+                        consumer.topic(),
+                        consumer.partition(),
+                        code.to_sentence(),
+                    );
+                    continue;
+                }
+                Err(other) => return Err(other.into()),
+            };
+
+            self.print_record(record.key(), record.value())?;
+        }
+
+        debug!("fetch loop exited");
+        println!("Consumer stream has closed");
+        Ok(())
+    }
+
+    /// Process fetch topic response based on output type
+    pub fn print_record(&self, key: Option<&[u8]>, value: &[u8]) -> Result<(), OutputError> {
```

Comment:
this should be pass as record? which should be more robust on more changes

Reply:
There is a specific reason I could not pass `fluvio::consumer::Record`, which is because the `fetch` interface from the consumer is still using the internal datatypes, not the `consumer::Record` data type. If we consolidated this, it would help to clean this up a little bit, but until then the common denominator is the bytes for key/value.

Reply:
reasonable